libvast/src/arrow_table_slice.cpp

```diff
@@ -1128,6 +1128,244 @@ arrow_table_slice<FlatBuffer>::record_batch() const noexcept {
+  VAST_ASSERT(batch->schema()->Equals(layout.to_arrow_schema()),
+              "VAST layout and Arrow schema must match");
+  VAST_ASSERT(std::is_sorted(transformations.begin(), transformations.end()),
+              "transformations must be sorted by index");
+  VAST_ASSERT(transformations.end()
+                == std::adjacent_find(
+                  transformations.begin(), transformations.end(),
+                  [](const auto& lhs, const auto& rhs) noexcept {
+                    const auto [lhs_mismatch, rhs_mismatch]
+                      = std::mismatch(lhs.index.begin(), lhs.index.end(),
+                                      rhs.index.begin(), rhs.index.end());
+                    return lhs_mismatch == lhs.index.end();
+                  }),
+              "transformation indices must not be a subset of the following "
+              "transformation's index");
+  // The current unpacked layer of the transformation, i.e., the pieces required
+  // to re-assemble the current layer of both the record type and the record
+  // batch.
+  struct unpacked_layer {
+    std::vector<struct record_type::field> fields;
+    arrow::ArrayVector arrays;
+  };
+  const auto impl
+    = [](const auto& impl, unpacked_layer layer, offset index, auto& current,
+         const auto sentinel) noexcept -> unpacked_layer {
+    VAST_ASSERT(!index.empty());
+    auto result = unpacked_layer{};
+    // Iterate over the current layer. For every entry in the current layer, we
+    // need to do one of three things:
+    // 1. Apply the transformation if the index matches the transformation
+    //    index.
+    // 2. Recurse to the next layer if the index is a prefix of the
+    //    transformation index.
+    // 3. Leave the elements untouched.
+    for (; index.back() < layer.fields.size(); ++index.back()) {
+      const auto [is_prefix_match, is_exact_match]
+        = [&]() noexcept -> std::pair<bool, bool> {
+        if (current == sentinel)
+          return {false, false};
+        const auto [index_mismatch, current_index_mismatch]
+          = std::mismatch(index.begin(), index.end(), current->index.begin(),
+                          current->index.end());
+        const auto is_prefix_match = index_mismatch == index.end();
+        const auto is_exact_match
+          = is_prefix_match && current_index_mismatch == current->index.end();
+        return {is_prefix_match, is_exact_match};
+      }();
+      if (is_exact_match) {
+        VAST_ASSERT(current != sentinel);
+        for (auto&& [field, array] : std::invoke(
+               std::move(current->fun), std::move(layer.fields[index.back()]),
+               std::move(layer.arrays[index.back()]))) {
+          result.fields.push_back(std::move(field));
+          result.arrays.push_back(std::move(array));
+        }
+        ++current;
+      } else if (is_prefix_match) {
+        auto nested_layer = unpacked_layer{
```

Comment:
How about an assert that we have a record_type?

Reply:
The `caf::get<record_type>(...)` internally asserts that already:

```cpp
template <vast::complex_type T, int Index>
static const T& caf::sum_type_access<vast::type>::get(
    const vast::type& x, sum_type_token<T, Index> token) {
  VAST_ASSERT(is(x, token));
  return static_cast<const T&>(
   static_cast<const vast::stateful_type_base&>(x));
}
```